### PR TITLE
Feat/query api client

### PR DESCRIPTION
**Pull request by baierjak**
## Query API client added

- ✨ useAppQuery for fetching data
- ✨ useAppMutation for mutating data on the BE side
- ✨ Hydration of data fetched on the server (getServerSideProps or getStaticProps)

based on #11

### CODE REVIEW

File: template/package.json [hunk 1]
```diff
@@ -25,9 +25,23 @@
         "snapshot-storybook": "npx chromatic --project-token=<passwd_token>"
     },
     "dependencies": {
+        "@ackee/antonio-core": "^4.0.14",
```
Comment (adamtretera) [inline template/package.json]: Hello there?

File: template/package.json [hunk 2]
```diff
@@ -25,9 +25,23 @@
...
+        "@storybook/addon-interactions": "^6.5.16",
+        "@storybook/addon-links": "^6.5.16",
+        "@storybook/builder-webpack5": "^6.5.16",
+        "@storybook/manager-webpack5": "^6.5.16",
+        "@storybook/react": "^6.5.16",
+        "@storybook/testing-library": "^0.1.0",
         "@svgr/cli": "^6.5.1",
+        "@tanstack/react-query": "^4.24.4",
+        "@tanstack/react-query-devtools": "^4.24.4",
+        "axios": "^1.3.2",
```
Comment (adamtretera) [inline template/package.json]: Delete

File: template/src/modules/api/services/handleError.ts [hunk 3]
```diff
@@ -0,0 +1,17 @@
+import { AxiosError } from 'axios';
```
Comment (adamtretera) [inline template/src/modules/api/services/handleError.ts]: delete

File: template/src/modules/api/types/strapi.ts [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+export interface StrapiResponse<Data> {
+    data: Data;
+    meta?: StrapiResponseMeta;
+    error?: StrapiResponseError;
+}
+
+export interface StrapiData<Data> {
+    id: number;
+    attributes: Data;
+}
```
Comment (adamtretera) [inline template/src/modules/api/types/strapi.ts]: Delete

File: template/src/pages/index.tsx [hunk 5]
```diff
@@ -126,86 +129,128 @@ export const button = css({
...
+    const { mutate } = useAppMutation(config.endpoints.testData, {
+        // Note: optional options
+    });
 
-                <div className={card()}>
-                    <a href="https://github.com/AckeeCZ/next-app-template/tree/main/docs/sentry.md">
-                        <h2>How to use Sentry &rarr;</h2>
-                        <p>Find out how to configure Sentry in the app and test the setup with button below.</p>
+    // eslint-disable-next-line no-console
+    console.log(data, isError, isLoading, isSuccess);
```
Comment (adamtretera) [inline template/src/pages/index.tsx:157]: console.log

File: template/src/modules/api/services/handleError.ts [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+    if (typeof err === 'string') {
+        error = { message: err.toUpperCase(), status: 500 };
+    } else if (err instanceof AxiosError) {
+        error = err;
+    } else {
+        error = { message: 'Error occured', status: 500 };
+    }
+
+    return error;
+}
```
Comment (cermakjiri) [inline template/src/modules/api/services/handleError.ts]: never used, can be removed

File: template/src/modules/api/services/mutationFnFactory.ts [hunk 7]
```diff
@@ -0,0 +1,16 @@
+import { config } from 'config';
+
+export const getMutationData = async <TPayload, TOptions>(endpoint: string, payload: TPayload, options?: TOptions) => {
+    const response = await fetch(config.api.url + endpoint, {
```
Comment (cermakjiri) [inline template/src/modules/api/services/mutationFnFactory.ts:4]:
```ts
new URL(endpoint, config.api.url)
```

File: template/src/modules/api/services/queryFnFactory.ts [hunk 8]
```diff
@@ -0,0 +1,15 @@
+import { config } from 'config';
+
+export const getQueryData = async <TOptions>(endpoint: string, options?: TOptions) => {
+    const response = await fetch(config.api.url + endpoint, {
```
Comment (cermakjiri) [inline template/src/modules/api/services/queryFnFactory.ts:4]:
```
new URL(endpoint, config.api.url)
```

File: template/src/modules/api/types/contact.ts [hunk 9]
```diff
@@ -0,0 +1,10 @@
+export interface ContactRequestPayload {
+    to?: string;
+    replyTo: string;
+    html: string;
+    subject: string;
+}
+
+export interface ContactData {
+    success: boolean;
+}
```
Comment (cermakjiri) [inline template/src/modules/api/types/contact.ts]: never used, pls remove

File: template/src/modules/api/types/index.ts [hunk 10]
```diff
@@ -0,0 +1,7 @@
+export * from './contact';
+export * from './strapi';
```
Comment (cermakjiri) [inline template/src/modules/api/types/index.ts]: delete

**Review by cermakjiri — CHANGES_REQUESTED** [review]

File: template/src/pages/index.tsx (hunk 5 above)
Comment (baierjak) [reply template/src/pages/index.tsx:157]: it is here on purpose to see the result from API, but good point and I will add a TODO to remove it

File: template/src/modules/api/services/queryFnFactory.ts (hunk 8 above)
Comment (baierjak) [reply template/src/modules/api/services/queryFnFactory.ts:4]: in #13

File: template/src/modules/api/services/mutationFnFactory.ts (hunk 7 above)
Comment (baierjak) [reply template/src/modules/api/services/mutationFnFactory.ts:4]: in #13

File: template/package.json [hunk 11]
```diff
@@ -25,6 +25,7 @@
         "snapshot-storybook": "npx chromatic --project-token=<passwd_token>"
     },
     "dependencies": {
+        "@babel/core": "^7.20.12",
```
Comment (jancama2) [inline template/package.json]: 👀

File: template/package.json [hunk 12]
```diff
@@ -33,6 +34,10 @@
         "react": "^18.2.0",
         "react-dom": "^18.2.0",
         "react-intl": "^6.0.5",
+        "@svgr/cli": "^6.5.1",
```
Comment (jancama2) [inline template/package.json]: 👀

File: template/package.json [hunk 13]
```diff
@@ -33,6 +34,10 @@
...
+        "@tanstack/react-query": "^4.24.4",
+        "@tanstack/react-query-devtools": "^4.24.4",
+        "babel-loader": "^8.3.0",
```
Comment (jancama2) [inline template/package.json]: 👀 the three packages here are not related to the PR, right?

**Review by jancama2 — APPROVED** [review]

**Review by adamtretera — APPROVED** [review]

File: template/package.json (hunk 11 above)
Comment (cermakjiri) [reply template/package.json]: i.e. what's it used for?

File: template/package.json (hunk 12 above)
Comment (cermakjiri) [reply template/package.json]: shouldn't be removed since `npx @svgr/cli` is used?

Comment (baierjak) [reply template/package.json]: yes but the previous changes wasnt pulled here automatically

**Pull request merged**